Demonstrating script address discovery and expose shared scripts in ApiWallet

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
https://jira.iohk.io/browse/ADP-525

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] Extend ApiWallet by sharedScripts
- [x] Updating testing layer of Api.Types
- [x] introducing `Metadata` having `TxMetadata` and `Script` and update code base
- [x] update tests wrt the above change
- [x] enable sending script via auxliaryScript in TxBody
- [x] discover scripts from SL.Tx in fromAllegra
- [x] use isShared in Model
- [x] add integration tests showing that the script with our verification keys is later discovered and exposed in sharedScripts in ApiWallet
- [ ] add more integration test with to wallets and script containing both verification keys
- [ ] deal with db migration

Prerequisite for using isShared:
- [x] PR merged -> https://github.com/input-output-hk/cardano-wallet/pull/2308 


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
absorbing shared script in ApiWallet

paweljakubas committed:
make the rest code compile and apply simplifications and corrections

paweljakubas committed:
add Arbitrary instances, tweak code to make unit tests in core pass

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -251,6 +251,13 @@ x-amount: &amount
+  enum:
+    - awaits_signing
+    - spent
+    - staked
```
I am not sure to understand the rationale behind those statuses? From our point of view, a script address exists on the blockchain or it doesn't. It'd be valuable however to show the balance available at that script address.

paweljakubas replied:
I expect to have the following lifecycle of addresses with script:
1. some funds are sent to this address
2. even more money can be sent
3. tx with script with all required signatures is sent
 and depending on where script is embedded, we have money spent or money staked
1-2 -> script awaits signing but available balance can be different
3 -> funds are unlocked and are spent or staked

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -1409,12 +1416,37 @@ components:
+        - funds_status
+        - funds_available
+      properties:
+        script_address: *AnyAddress
```
Seems like there's some excessive wrapping here. The resulting JSON will look like `{ "script_address": { "address": "..." } }`

What about simply showing it as `"address": ...` ?

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -1409,12 +1416,37 @@ components:
+          items: *ApiVerificationKey
+        funds_status: *scriptStatus
+        funds_available: *amount
+
```
I think it'll be valuable to show a JSON representation of the whole script no?

paweljakubas committed:
update schema

make TypeSpec unit tests pass

remove ApiScriptStatus

regenerate jsons

try to make jsons smaller

make Script generator smaller

revert making script gen smaller

KtorZ commented on lib/core/test/data/Cardano/Wallet/Api/ApiWallet.json:
```diff
@@ -1,476 +1,15262 @@
 {
-    "seed": 287889657345073705,
+    "seed": -7459132351643344646,
```
That file is _way_ too large :grimacing:

paweljakubas committed:
introduce Metadata consisting of TxMetada and Script and propagate it through codebase

paweljakubas committed:
update unit testing

hlint + stylish

paweljakubas committed:
add proper ToJSON/FromJSON for Metadata

paweljakubas committed:
align schema and testing

paweljakubas committed:
fix CLISpec

paweljakubas committed:
support toCardanoScript conversion and use it in auxScript of TxBody from node

paweljakubas committed:
revert one change to have all integration tests passing

paweljakubas committed:
add integration test sending script in auxScript with FeeTooSmallUTxO

paweljakubas committed:
account script in estimateMinimumFee in Shelley.Transaction

paweljakubas committed:
move forward with reaading scripts from ledger tx

fix constraints

Commit:
Regenerate nix

paweljakubas committed:
add persistent to MetaScript and demonstrate that discovery of tx with script works

paweljakubas committed:
try adding isShared to model

first full integration test working

update core unit tests

post rebase fix
